fuel/datasets/hdf5.py

```diff
@@ -495,87 +467,62 @@ def load(self):
-                    if source_name in split_indices:
-                        ind = split_indices[source_name]
-                    else:
-                        ind = list(range(*split_start_stop[source_name]))
-                    indices[source_name] = sorted(
-                        set(indices[source_name] + ind))
-        else:
-            start_stop = self.get_start_stop(handle, self.which_sets[0])
-            indices = self.get_indices(handle, self.which_sets[0])
-        num_examples = None
-        for source_name in self.sources:
-            subset = self._subset_template
-            # If subset has a step greater than 1, we convert it to a list,
-            # otherwise we won't be able to take that subset within the context
-            # of a split defined by a slice.
-            if hasattr(subset, 'step') and subset.step not in (1, None):
-                subset = list(range(len(handle[source_name])))[subset]
-            self._subset_template = subset
-            if source_name in indices:
-                source_subset = indices[source_name]
-            else:
-                source_subset = slice(*start_stop[source_name])
-            if hasattr(subset, 'step') and hasattr(source_subset, 'step'):
-                subset = slice(
-                    source_subset.start
-                    if subset.start is None else subset.start,
-                    source_subset.stop
-                    if subset.stop is None else subset.stop,
-                    subset.step)
-                subsets.append(subset)
-                subset_num_examples = subset.stop - subset.start
-            else:
-                if hasattr(source_subset, 'step'):
-                    source_subset = numpy.arange(
-                        source_subset.start, source_subset.stop)
-                subset = source_subset[subset]
-                subsets.append(subset)
-                subset_num_examples = len(subset)
-            if num_examples is None:
-                num_examples = subset_num_examples
-            if num_examples != subset_num_examples:
-                raise ValueError("sources have different lengths")
-        self.subsets = subsets
-
-        # Load data sources and source shapes
+        # Infer subsets based on `which_sets`
+        subsets = [None for source_name in self.sources]
+        for split in self.which_sets:
+            start_stop = self.get_start_stop(handle, split)
+            indices = self.get_indices(handle, split)
+            for i, source_name in enumerate(self.sources):
+                if source_name in indices:
+                    source_split_subset = Subset(
+                        indices[source_name], len(handle[source_name]))
+                else:
+                    source_split_subset = Subset(
+                        slice(*start_stop[source_name]),
+                        len(handle[source_name]))
+                if subsets[i] is None:
```

Comment:
Actually, there are already safegards in place that check whether we're trying to request something within the context of a subset. They would raise an error because the request would always be greater than subset's number of examples (0).

I revise my position and accept your suggestion.